sync/blocksync/syncer_test.go

```diff
@@ -0,0 +1,319 @@
+			expectedBlocks:           []int{3, 4, 5},
+			verifyZeroBlocksReceived: true,
+		},
+		{
+			name:              "some blocks already available",
+			numBlocks:         10,
+			prePopulateBlocks: []int{4, 5},
+			fromHeight:        5,
+			parentsToGet:      3,
+			expectedBlocks:    []int{3, 4, 5},
+		},
+		{
+			name:              "most recent block missing",
+			numBlocks:         10,
+			prePopulateBlocks: []int{3, 4},
+			fromHeight:        5,
+			parentsToGet:      3,
+			expectedBlocks:    []int{3, 4, 5},
+		},
+		{
+			name:           "edge case - from height 1",
+			numBlocks:      10,
+			fromHeight:     1,
+			parentsToGet:   1,
+			expectedBlocks: []int{1},
+		},
+		{
+			name:           "single block sync",
+			numBlocks:      10,
+			fromHeight:     7,
+			parentsToGet:   1,
+			expectedBlocks: []int{7},
+		},
+		{
+			name:           "large sync - many blocks",
+			numBlocks:      50,
+			fromHeight:     40,
+			parentsToGet:   35,
+			expectedBlocks: []int{6, 10, 20, 30, 40},
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			env := newTestEnvironment(t, tt.numBlocks).withDefaultBlockProvider()
+
+			if len(tt.prePopulateBlocks) > 0 {
+				env.prePopulateBlocks(tt.prePopulateBlocks...)
+			}
+
+			syncer, err := env.createSyncer(tt.fromHeight, tt.parentsToGet)
+			require.NoError(t, err)
+
+			ctx := context.Background()
+			require.NoError(t, syncer.Start(ctx))
+			require.NoError(t, syncer.Wait(ctx))
+
+			env.verifyBlocksInDB(t, tt.expectedBlocks)
+
+			if tt.verifyZeroBlocksReceived {
+				// Client should not have received any block requests since all blocks were on disk
+				require.Equal(t, int32(0), env.client.BlocksReceived())
+			}
+		})
+	}
+}
+
+func TestBlockSyncer_ContextCancellation(t *testing.T) {
+	// Allow some time for startup for cancellation
+	env := newTestEnvironment(t, 10).withCustomBlockProvider(func(hash common.Hash, height uint64) *types.Block {
+		time.Sleep(100 * time.Millisecond)
+		return nil
+	})
+
+	syncer, err := env.createSyncer(5, 3)
+	require.NoError(t, err)
+
+	ctx, cancel := context.WithCancel(context.Background())
+	require.NoError(t, syncer.Start(ctx))
+
+	// Cancel context immediately
+	cancel()
+
+	err = syncer.Wait(ctx)
+	require.ErrorIs(t, err, context.Canceled)
+}
+
+// testEnvironment provides an abstraction for setting up block syncer tests
+type testEnvironment struct {
+	chainDB ethdb.Database
+	client  *syncclient.TestClient
+	blocks  []*types.Block
+}
+
+// newTestEnvironment creates a new test environment with generated blocks
+func newTestEnvironment(t *testing.T, numBlocks int) *testEnvironment {
+	t.Helper()
+
+	var (
+		key1, _        = crypto.HexToECDSA("b71c71a67e1177ad4e901695e1b4b9ee17ae16c6668d313eac2f96dbcda3f291")
+		key2, _        = crypto.HexToECDSA("8a1f9a8f95be41cd7ccb6168179afb4504aefe388d1e14474d32c45c72ce7b7a")
+		addr1          = crypto.PubkeyToAddress(key1.PublicKey)
+		addr2          = crypto.PubkeyToAddress(key2.PublicKey)
```

Comment:
`key2` and `addr2` are really not needed. Any place that uses `addr2` could just use `addr1`.